plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/ui/internal/view/CDKActionProvider.java

```diff
@@ -0,0 +1,261 @@
+					// If this is the server that was / is selected
+					if( previousSelection != null && previousSelection.size() > 0 
+							&& previousSelection.getFirstElement().equals(event.getServer())) {
+						// and it switches state, update enablement
+						Display.getDefault().asyncExec(new Runnable() {
+							public void run() {
+								if( UnitedServerListener.serverSwitchesToState(event, IServer.STATE_STARTED)) {
+									setEnabled(true);
+								} else if( UnitedServerListener.serverSwitchesToState(event, IServer.STATE_STOPPED)) {
+									setEnabled(false);
+								}
+							}
+						});
+					}
+				}
+			};
+			selectionChanged(sp.getSelection());
+		}
+		
+		protected Object adaptToViewItem(IServer server) {
+			// Subclass override this
+			return null;
+		}
+		
+		public boolean accept(IServer server) {
+			boolean preconditions = (server.getServerType() != null && 
+					adaptToViewItem(server) != null
+					&& server.getServerState() == IServer.STATE_STARTED);
+			return preconditions;
+		}
+		
+		public void selectionChanged(IStructuredSelection sel) {
+			if( sel.size() != 1 ) {
+				setEnabled(false);
+				return;
+			}
+			setEnabled(true);
+			synchronized(this) {
+				switchListener(previousSelection, sel);
+				previousSelection = sel;
+			}
+			super.selectionChanged(sel);
+		}
+
+		private void switchListener(IStructuredSelection previousSelection, IStructuredSelection newSel) {
+			if( previousSelection != null ) {
+				Object o = previousSelection.getFirstElement();
+				if( o instanceof IServer) {
+					((IServer)o).removeServerListener(serverListener);
+				}
+			}
+			Object newSel1 = newSel.getFirstElement();
+			if( newSel1 instanceof IServer) {
+				((IServer)newSel1).addServerListener(serverListener);
+			}
+		}
+		
+		public void perform(final IServer server) {
+			// Only run in UI thread
+			
+			IWorkbenchPart part = null;
+			try {
+				part = bringViewToFront(viewId);
+			} catch(PartInitException pie) {
+				CDKCoreActivator.pluginLog().logError("Error opening view " + viewId, pie);
+			}
+			
+			if( part != null ) {
+				final CommonNavigator view = (CommonNavigator) part.getAdapter(CommonNavigator.class);
+				if (view != null && view.getCommonViewer() != null && view.getCommonViewer().getTree() != null && !view.getCommonViewer().getTree().isDisposed()){
+					Object connection = adaptToViewItem(server);
+					if( connection != null ) {
+						view.getCommonViewer().collapseAll();
+						view.getCommonViewer().expandToLevel(2);
+						ISelection sel = new StructuredSelection(new Object[] { connection });
+						view.getCommonViewer().setSelection(sel, true);
+					}
+				}
+			}
+		}
+	}
+	private static final IWorkbenchPart bringViewToFront(String viewId) throws PartInitException {
+		IWorkbenchWindow window = PlatformUI.getWorkbench().getActiveWorkbenchWindow() ;
+		IWorkbenchPart part = null;
+		if (window != null) {
+			IWorkbenchPage page = window.getActivePage();
+			if (page != null) {
```

Comment:
Why just part = page.showView(viewId); is not enough?
There are about 20 places in JBossTools that use showView() with similar goal - activate and focus a view part. If this algorithm solves some problem it should be shared for all those calls.
There are now 3 copies of bringViewToFront() methods:
1) org.jboss.ide.eclipse.as.ui.UIUtil - it is marked as deprrecated, but nevertheless is used in openshift's common ConsoleUtils.
2) JMXProvider - as private.
3) Here - as private.